### PR TITLE
Utils: Code Revision - Update network common utils

### PR DESCRIPTION
**Pull request by yukirora**
**Description**
Update network common utils.

**Major Revision**
- Add get_ib_devices in network common utils 
- Move get_free_port from test utils to network common utils

### CODE REVIEW

File: superbench/common/utils/network.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    """Get a free port in current system.
+
+    Return:
+        port (int): a free port in current system.
+    """
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind(('', 0))
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return s.getsockname()[1]
+
```
Comment (cp5555) [inline superbench/common/utils/network.py:21]: What will we return if closing fails?

File: superbench/common/utils/network.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from contextlib import closing
+
+
+def get_free_port():
+    """Get a free port in current system.
+
+    Return:
+        port (int): a free port in current system.
+    """
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
```
Comment (cp5555) [inline superbench/common/utils/network.py:17]: What's the benefit of using `with closing()`?

File: superbench/common/utils/network.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    # Filter 'InfiniBand' devices by link_layer
+    command = "ibv_devinfo | awk '$1 ~ /hca_id/||/link_layer:/ {print $1,$2}'"
+    output = subprocess.run(
+        command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True, check=False, universal_newlines=True
+    )
+    lines = list(filter(None, output.stdout.replace('\n', ' ').split('hca_id:')))
+    ib_devices = []
+    for line in lines:
+        if 'InfiniBand' in line:
```
Comment (cp5555) [inline superbench/common/utils/network.py:37]: I think you should check `transport: InfiniBand`.

File: superbench/common/utils/network.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+def get_ib_devices():
+    """Get available IB devices in the system and filter ethernet devices.
+
+    Return:
+        ib_devices (list): IB devices in current system.
+    """
+    # Filter 'InfiniBand' devices by link_layer
+    command = "ibv_devinfo | awk '$1 ~ /hca_id/||/link_layer:/ {print $1,$2}'"
```
Comment (cp5555) [inline superbench/common/utils/network.py:30]: Why not `ibv_devinfo` directly?

File: superbench/common/utils/network.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    output = subprocess.run(
+        command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True, check=False, universal_newlines=True
+    )
```
Comment (cp5555) [inline superbench/common/utils/network.py:33]: What happens if we don't install `ibv_devinfo`?

File: superbench/common/utils/__init__.py [hunk 6]
```diff
@@ -10,11 +10,6 @@
...
 __all__ = [
-    'LazyImport',
-    'SuperBenchLogger',
-    'create_sb_output_dir',
-    'get_sb_config',
-    'logger',
-    'nv_helper',
-    'rotate_dir',
+    'LazyImport', 'SuperBenchLogger', 'create_sb_output_dir', 'get_sb_config', 'logger', 'nv_helper', 'rotate_dir',
+    'network'
```
Comment (abuccts) [inline superbench/common/utils/__init__.py:14]: add trailing comma
sort in alphabet order

File: superbench/common/utils/network.py (hunk 4 above)
Comment (abuccts) [inline superbench/common/utils/network.py:30]: check `/sys/class/infiniband/<device>/ports/<port-num>/link_layer` is `InifiniBand` and get `<device>` list?

reference: [sysfs-class-infiniband](https://www.kernel.org/doc/Documentation/ABI/stable/sysfs-class-infiniband)

Comment (abuccts) [reply superbench/common/utils/network.py:30]: could also check `/sys/class/infiniband/<device>/node_type` for transport type
referemce: [dev->transport_type](https://github.com/linux-rdma/rdma-core/blob/d7a9dde2b164a1563c7a51e4ade8ea3b6cdea9ef/libibverbs/init.c#L380-L401)

File: superbench/common/utils/__init__.py [hunk 7]
```diff
@@ -10,11 +10,6 @@
...
 
 __all__ = [
-    'LazyImport',
-    'SuperBenchLogger',
-    'create_sb_output_dir',
-    'get_sb_config',
-    'logger',
-    'nv_helper',
-    'rotate_dir',
+    'LazyImport', 'SuperBenchLogger', 'create_sb_output_dir', 'get_sb_config', 'logger', 'nv_helper', 'rotate_dir',
```
[the remark this answers is not in the document]
Comment (yukirora) [reply superbench/common/utils/__init__.py]: It was automatically changed by lint and format, not me

File: superbench/common/utils/network.py (hunk 1 above)
Comment (yukirora) [reply superbench/common/utils/network.py:25]: change to try-catch style, but close() was designed to put in finally

File: superbench/common/utils/network.py (hunk 2 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: it's an error-handling for socket exception

File: superbench/common/utils/network.py (hunk 3 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: transport can not differentiate

File: superbench/common/utils/network.py (hunk 4 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: change to another check interface according to Yifan's comments

File: superbench/common/utils/network.py (hunk 5 above)
Comment (yukirora) [reply superbench/common/utils/network.py:41]: change to another check interface according to Yifan's comments

File: superbench/common/utils/__init__.py (hunk 6 above)
Comment (yukirora) [reply superbench/common/utils/__init__.py]: thank you

File: superbench/common/utils/__init__.py [hunk 8]
```diff
@@ -10,6 +10,6 @@
 nv_helper = LazyImport('superbench.common.utils.nvidia_helper')
 
 __all__ = [
-    'LazyImport', 'SuperBenchLogger', 'create_sb_output_dir', 'get_sb_config', 'logger', 'nv_helper', 'rotate_dir',
-    'network'
+    'LazyImport', 'SuperBenchLogger', 'create_sb_output_dir', 'get_sb_config', 'logger', 'network', 'nv_helper',
+    'rotate_dir'
```
Comment (cp5555) [inline superbench/common/utils/__init__.py:14]: one item for one line?

**Review by cp5555 — APPROVED** [review]

File: superbench/common/utils/__init__.py (hunk 8 above)
Comment (guoshzhao) [reply superbench/common/utils/__init__.py:14]: For benchmarks submodule, the items are placed in one line.

**Review by guoshzhao — APPROVED** [review]

File: superbench/common/utils/__init__.py (hunk 8 above)
Comment (abuccts) [reply superbench/common/utils/__init__.py:14]: it will exceed max line length

File: superbench/common/utils/__init__.py [hunk 9]
```diff
@@ -10,11 +10,6 @@
...
 __all__ = [
-    'LazyImport',
-    'SuperBenchLogger',
-    'create_sb_output_dir',
-    'get_sb_config',
-    'logger',
-    'nv_helper',
-    'rotate_dir',
+    'LazyImport', 'SuperBenchLogger', 'create_sb_output_dir', 'get_sb_config', 'logger', 'network', 'nv_helper',
+    'rotate_dir'
```
Comment (abuccts) [inline superbench/common/utils/__init__.py:14]:
```suggestion
    'rotate_dir',
```

File: superbench/common/utils/network.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+
+def get_ib_devices():
+    """Get available IB devices in the system and filter ethernet devices.
+
+    Return:
+        ib_devices (list): IB devices in current system.
+    """
+    command_get_devices = 'ls /sys/class/infiniband/'
+    output = subprocess.run(
```
Comment (abuccts) [inline superbench/common/utils/network.py:34]: use file operations instead of subprocess?

File: superbench/common/utils/__init__.py (hunk 8 above)
Comment (guoshzhao) [reply superbench/common/utils/__init__.py:14]: Why we should concern about this rule? New line can be created if exceed the max-line-length, right? Please refer:

superbench/benchmarks/micro_benchmarks/__init__.py

File: superbench/common/utils/network.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        command_get_devices,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        shell=True,
+        check=False,
+        universal_newlines=True
+    )
+    devices = output.stdout.split()
+    devices.sort()
```
Comment (abuccts) [inline superbench/common/utils/network.py:43]: use natural sort order? e.g., what if there are 16 devices mlx5_0, mlx5_1, ..., mlx5_15
```suggestion
    devices.sort(key=lambda s: [int(ch) if ch.isdigit() else ch for ch in re.split(r'(\d+)', s)])
```

File: superbench/common/utils/network.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            output = subprocess.run(
+                command_get_link_layer,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                shell=True,
+                check=False,
+                universal_newlines=True
+            )
+            if 'InfiniBand' in output.stdout:
+                ib_devices.append(device)
```
Comment (abuccts) [inline superbench/common/utils/network.py:68]: what if one hca has two ports? here will append it twice

File: superbench/common/utils/__init__.py (hunk 9 above)
Comment (yukirora) [reply superbench/common/utils/__init__.py]: thank you for the suggestions

File: superbench/common/utils/network.py (hunk 10 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: ok, thank you very much

File: superbench/common/utils/network.py (hunk 11 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: thanks for the comment

File: superbench/common/utils/network.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        return None
+    finally:
+        s.close()
+
+
+def get_ib_devices():
+    """Get available IB devices with available ports in the system and filter ethernet devices.
+
+    Return:
+        ib_devices (list): IB devices with available ports in current system.
```
Comment (abuccts) [inline superbench/common/utils/network.py:32]: update accordingly?

**Review by abuccts — APPROVED** [review]

File: superbench/common/utils/network.py (hunk 13 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: updated

File: superbench/common/utils/network.py (hunk 4 above)
Comment (yukirora) [reply superbench/common/utils/network.py]: thank you very much

File: superbench/common/utils/__init__.py (hunk 8 above)
Comment (abuccts) [reply superbench/common/utils/__init__.py]: this list is updated before when it cannot be fit in one line.
isn't it more clear to have one item per line (adding a trailing comma for last item could achieve this) and sort in alphabet order if the list is too long? e.g., [when it has 50+ items](https://github.com/pytorch/pytorch/blob/v1.9.0/torch/distributions/__init__.py#L118)

Comment (guoshzhao) [reply superbench/common/utils/__init__.py]: If you have these kind of concerns, we should make a rule to let all list of item to be placed in format 'one item per line', including function arguments, list, tupe and so on.
For __all__ in __init__.py, it's just to control the exporting, not for human readable.

https://github.com/python/cpython/blob/main/Lib/logging/__init__.py#L32

Comment (abuccts) [reply superbench/common/utils/__init__.py]: different module in one project may have different styles, so nobody uses a lint rule to force `__all__`
https://github.com/python/cpython/blob/3.10/Lib/concurrent/futures/__init__.py#L20

**Pull request merged**